rustc_ast: Visit tokens stored in AST nodes in mutable visitor

petrochenkov opened the pull request:
After #77271 token visiting is enabled only for one visitor in `rustc_expand\src\mbe\transcribe.rs` which applies hygiene marks to tokens produced by declarative macros (`macro_rules` or `macro`), so this change doesn't affect anything else.

When a macro has some interpolated token from an outer macro in its output
```rust
macro inner() {
    $interpolated
}
```
we can use the usual interpretation of interpolated tokens in token-based model - a None-delimited group - to write this macro in an equivalent form
```rust
macro inner() {
    ⟪ a b c d ⟫
}
```

When we are expanding the macro `inner` we need to apply hygiene marks to all tokens produced by it, including the tokens inside the group.

Before this PR we did this by visiting the AST piece inside the interpolated token and applying marks to all spans in it.
I'm not sure this is 100% correct (ideally we should apply the marks to tokens and then re-parse the AST from tokens), but it's a very good approximation at least.
We didn't however apply the marks to actual tokens stored in the nonterminal, so if we used the nonterminal as a token rather than as an AST piece (e.g. passed it to a proc macro), then we got hygiene bugs.
This PR applies the marks to tokens in addition to the AST pieces thus fixing the issue.

r? @Aaron1011

petrochenkov committed:
rustc_ast: Visit tokens stored in AST nodes in mutable visitor

petrochenkov committed:
rustc_ast: Never clone empty token streams in mutable visitor

Aaron1011 commented:
> I'm not sure this is 100% correct (ideally we should apply the marks to tokens and then re-parse the AST from tokens), but it's a very good approximation at least.

That would just affect the computed spans, right? E.g. we might fail to join two spans that now have different `SyntaxContexts`?

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit fcfe3063cbacb070da03eeda8ed88c7bfd24f9de with merge bd450b36446fdb9d6276544cc56174298a890781...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fcfe3063cbacb070da03eeda8ed88c7bfd24f9de","merge_sha":"bd450b36446fdb9d6276544cc56174298a890781"} -->

Aaron1011 commented on src/test/ui/proc-macro/nonterminal-token-hygiene.rs:
```diff
@@ -0,0 +1,27 @@
+macro_rules! outer {
+    ($item:item) => {
+        macro inner() {
+            recollect! { $item }
```
Could you change this to use `print_bang`, and make this test file like https://github.com/rust-lang/rust/blob/0fb00251a578018430bad6687ff74e14031d2a07/src/test/ui/proc-macro/meta-macro-hygiene.rs (e.g. `-Z unpretty=expanded,hygiene` and other flags)? This behavior is quite subtle, and I think it would be good to ensure it never accidentally changes.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1356556448)
Build commit: bd450b36446fdb9d6276544cc56174298a890781 (`bd450b36446fdb9d6276544cc56174298a890781`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1356556448"},"merge_sha":"bd450b36446fdb9d6276544cc56174298a890781"} -->

rust-timer commented:
Queued bd450b36446fdb9d6276544cc56174298a890781 with parent 0fb00251a578018430bad6687ff74e14031d2a07, future [comparison URL](https://perf.rust-lang.org/compare.html?start=0fb00251a578018430bad6687ff74e14031d2a07&end=bd450b36446fdb9d6276544cc56174298a890781).

rust-timer commented:
Finished benchmarking try commit (bd450b36446fdb9d6276544cc56174298a890781): [comparison url](https://perf.rust-lang.org/compare.html?start=0fb00251a578018430bad6687ff74e14031d2a07&end=bd450b36446fdb9d6276544cc56174298a890781).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-perf

petrochenkov commented:
@Aaron1011 
>That would just affect the computed spans, right? E.g. we might fail to join two spans that now have different `SyntaxContexts`?

Yes, perhaps some spans will be different.
Perhaps it's possible to make an example where edition-specific parsing would be different and produce an error (but that seems to be because our edition hygiene works incorrectly with nested macro definitions - the outermost `ExpnData` is used instead of the innermost one).

petrochenkov committed:
More detailed output for the nonterminal marking test

petrochenkov commented:
Updated.

I've also added the mut visitor micro-optimization commit from https://github.com/rust-lang/rust/pull/78736 which seems relevant and non-controversial.

Aaron1011 commented:
@bors r+

bors commented:
:pushpin: Commit 8def2fc122a9f698914606c51a9e8c18aed168df has been approved by `Aaron1011`

<!-- @bors r=Aaron1011 8def2fc122a9f698914606c51a9e8c18aed168df -->
<!-- homu: {"type":"Approved","sha":"8def2fc122a9f698914606c51a9e8c18aed168df","approver":"Aaron1011"} -->

bors commented:
:hourglass: Testing commit 8def2fc122a9f698914606c51a9e8c18aed168df with merge 1773f60ea5d42e86b8fdf78d2fc5221ead222bc1...
<!-- homu: {"type":"BuildStarted","head_sha":"8def2fc122a9f698914606c51a9e8c18aed168df","merge_sha":"1773f60ea5d42e86b8fdf78d2fc5221ead222bc1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1371393638)
Approved by: Aaron1011
Pushing 1773f60ea5d42e86b8fdf78d2fc5221ead222bc1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Aaron1011","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1371393638"},"merge_sha":"1773f60ea5d42e86b8fdf78d2fc5221ead222bc1"} -->
